src/main/java/com/andrewthesailor/cinemaTickets/service/impl/ReservationServiceImpl.java

```diff
@@ -0,0 +1,156 @@
+            throw new TicketException("Reservation not possible: reservations are closed");
+        }
+    }
+
+    private void verifySeats(ReservationRequestData data, Screening screening) throws TicketException {
+        List<Seat> freeSeats = seatService.getFreeSeats(screening);
+        List<Long> ids = data.getTicketList().stream().flatMap(p -> Stream.of(p.getSeatId())).collect(Collectors.toList());
+        List<String> ticketTypeCodes = data.getTicketList().stream().flatMap(p -> Stream.of(p.getTicketTypeCode())).collect(Collectors.toList());
+        findDuplicates(ids);
+        findWrongSeats(ids, freeSeats);
+        verifySeatOrder(ids, screening);
+        verifyNameAndSurname(data.getName(), data.getSurname());
+        ticketTypeService.verifyCodes(ticketTypeCodes);
+    }
+
+    private void verifyNameAndSurname(String name, String surname) {
+        if (!stringUtil.verifyName(name)) {
+            throw new TicketException("Invalid name");
+        }
+        if (!stringUtil.verifySurname(surname)) {
+            throw new TicketException("Invalid surname");
+        }
+
+    }
+
+    private void verifySeatOrder(List<Long> ids, Screening screening) {
+        List<Seat> seatsToReserve = seatService.getSeats(ids);
+        List<Seat> reservedSeats = seatService.getReservedSeats(screening);
+        List<Integer> rowsToVerify = seatsToReserve.stream().flatMap(p -> Stream.of(p.getRow())).distinct().collect(Collectors.toList());
+        List<Integer> invalidRows = new ArrayList<>();
+        for (Integer row : rowsToVerify) {
+            List<Seat> seats = seatsToReserve.stream().filter(p -> p.getRow().equals(row)).collect(Collectors.toList());
+            seats.addAll(reservedSeats.stream().filter(p -> p.getRow().equals(row)).collect(Collectors.toList()));
+            seats.sort(Comparator.comparingInt(Seat::getColumn));
+            Integer seatSize = seats.size() - 1;
+            Integer reservedSeatCount = seats.get(seatSize).getColumn() - seats.get(0).getColumn();
+            if (!seatSize.equals(reservedSeatCount)) {
+                invalidRows.add(row);
+            }
+        }
+        if (!invalidRows.isEmpty()) {
+            throw new TicketException("Seats in rows: " + invalidRows + " are leaving empty spaces");
+        }
+    }
+
+    private void findDuplicates(List<Long> ids) {
+        List<Long> duplicates = ids.stream().collect(Collectors.groupingBy(Function.identity(), Collectors.counting()))
+                .entrySet().stream().filter(i -> i.getValue() > 1).map(Map.Entry::getKey).collect(Collectors.toList());
+        if (!duplicates.isEmpty())
+            throw new TicketException("Duplicate seats with id:" + duplicates);
+    }
+
+    private void findWrongSeats(List<Long> ids, List<Seat> freeSeats) {
+        List<Long> seatsIds = freeSeats.stream().flatMap(p -> Stream.of(p.getId())).collect(Collectors.toList());
+        List<Long> wrongSeats = new ArrayList<>();
```

Comment:
`wrong` to mało mówiąca nazwa.

Reply:
invalid/conflicting pewnie byłoby lepsze, w rzeczywistej apce pewnie rozbiłbym ten check na dwa- siedzenie nie istnieje/jest zajęte